improved array handling in cli

fishpepper opened the pull request:
I will need better cli array support before I can continue with #3417.

Therefore I added support for uint8_t, int8_t, uint16_t, and int16_t arrays.
You can define such arrays e.g. by:
`  "array_unit_test",             VAR_INT8  | MODE_ARRAY | MASTER_VALUE, .config.array.length = 3, PG_RESERVED_FOR_TESTING_1, 0 
`

I also reworked the parser a bit (should be more efficient like this, no need to copy substrings etc).

I also added a basic unit test for the array support. Due to missing array examples in the settings.c file I added the beforementioned array_unit_test entry.
This can be removed from settings.c once we have entries that use this feature.
The unit test could probably be cleaner/with less dummy stubs by more clever inclusion of files but it does its job.

The whole thing is not tested on a real FC right now.

DanNixon commented on src/test/unit/unittest_cli.h:
```diff
@@ -0,0 +1,26 @@
+/*
```
This file seems unnecessary.

DanNixon commented:
A better approach to the unit test would probably be to not compile `settings.c` and simply define a `valueTable[]` that is unit test specific.

This keeps unit test specific code out of the firmware code and avoids the need for all the PG declarations and function stubs in the test.

fishpepper commented:
Good idea! 
In my local branch I have my osd stuff checked out as well. I was using/testing the osd array stuff with the unittest. I removed the stuff for this pull request and added the synthetic test case in settings.c.

I added the valueTable and some more stubs from settings.c - works fine :)
I did some experiments removing PG defines and stubs. But it explodes again due to missing references. So for now I kept them in the unittest.

DanNixon commented:
Ah, yeah. A lot of thost stubs will be required for `cli.c` includes too...

martinbudden commented on src/main/fc/cli.c:
```diff
@@ -302,9 +302,36 @@ static void printValuePointer(const clivalue_t *var, const void *valuePointer, b
-            uint8_t value = ((uint8_t *)valuePointer)[i];
+            switch (var->type & VALUE_TYPE_MASK) {
+                default:
+                case(VAR_UINT8): {
```
Some minor coding style points. BF style is to align `case` with `switch`, and have no brackets around case item. Also braces within `case` statements are indented and below `case` keyword, but are better avoided if possible. So
```
    switch (var->type & VALUE_TYPE_MASK) {
    default:
    case VAR_UINT8:
        // fetch data
        cliPrintf("%d", ((uint8_t *)valuePointer)[i]);
        break;
```

martinbudden commented:
I concur with @DanNixon , if you declare (eg) `PG_REGISTER` rather than `PG_REGISTER_WITH_RESET_FN` you avoid the need to create dummy reset functions.

martinbudden commented on src/test/unit/cli_unittest.cc:
```diff
@@ -0,0 +1,306 @@
+uint32_t getBeeperOffMask(void) { return 0; }
+uint32_t getPreferredBeeperOffMask(void) { return 0; }
+
+void beeper(beeperMode_e mode) {UNUSED(mode);}
```
Generally style for stub functions is (eg):
```
void beeper(beeperMode_e) {}
```
By omitting the variable name you avoid the need for `UNUSED`.

fishpepper commented on src/main/fc/cli.c:
```diff
@@ -2772,38 +2766,38 @@ STATIC_UNIT_TESTED void cliSet(char *cmdline)
+                                    // fetch data pointer
+                                    uint8_t *data = (uint8_t *)getValuePointer(val) + i;
+                                    // store value
+                                    *data = (uint8_t)atoi((const char*) valPtr);
```
Ok, I included the suggested changes.

Getting rid of the brackets on these switch cases makes it quite unreadable.
An option would be to have the different datatype pointers defined at the head of the function like
uint8_t *uint8_data_ptr;
int8_t *int8_data_ptr;
...

martinbudden replied:
I'm not against using brackets in `switch` statements, just against using them when they can be easily avoided.  So I wouldn't get rid of them here.
